A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,207 @@
+----
+* Author(s): murgatroid99
+* Approver: markdroth
+* Implemented in: C++, Java, Go, Node
+* Last updated: 2021-11-30
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+gRPC will add support for the xDS [Outlier Detection feature](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier). This feature allows users to configure heuristics that clients can use to decide to stop sending requests to a specific backend because it appears unhealthy.
+
+## Background
+
+Service mesh orchestration systems perform active health checking and provide that information to clients. Outlier Detection is an additional heuristic-based passive health checking strategy that clients can perform directly.
+
+
+### Related Proposals:
+
+This proposal builds on earlier work described in the following gRFCs:
+* [gRFC A27: xDS-based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+* [gRFC A30: xDS v3 Support](https://github.com/grpc/proposal/blob/master/A30-xds-v3.md)
+* [gRFC A37: xDS Aggregate and Logical DNS Clusters](https://github.com/grpc/proposal/blob/master/A37-xds-aggregate-and-logical-dns-clusters.md)
+* [gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md)
+
+## Proposal
+
+We will add a new `outlier_detection` load balancing policy that implements the outlier detection algorithm. This load balancing policy itself is not specific to xDS, but the xDS LB policy tree will use it.
+
+### LB Policy Config
+
+The `outlier_detection` policy will have the following config. All fields except the last one come directly from Envoy's [`Outlier Detection` message](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/cluster/v3/outlier_detection.proto), and the comments are modified from the comments there:
+
+```proto
+message OutlierDetectionLoadBalancingConfig {
+  // The time interval between ejection analysis sweeps. This can result in
+  // both new ejections as well as addresses being returned to service. Defaults
+  // to 10000ms or 10s.
+  google.protobuf.Duration interval = 1;
+
+  // The base time that as address is ejected for. The real time is equal to the
+  // base time multiplied by the number of times the address has been ejected.
+  // Defaults to 30000ms or 30s.
+  google.protobuf.Duration base_ejection_time = 2;
+
+  // The maximum time that an address is ejected for. If not specified, the default value (300000ms or 300s) or
+  // the base_ejection_time value is applied, whatever is larger.
+  google.protobuf.Duration max_ejection_time = 3;
+
+  // The maximum % of an address list that can be ejected due to outlier
+  // detection. Defaults to 10% but will eject at least one address regardless of the value.
+  google.protobuf.UInt32Value max_ejection_percent = 4;
+
+  // This factor is used to determine the ejection threshold for success rate
```

Comment:
One thing to consider here is that the current design accepts a configuration object that is empty other than the `child_policy`, which will result in all fields having their default values. Notably, that will effectively enable `success_rate` ejection and disable `failure_percentage`, because the default `enforcing_success_rate` is 100 and the default `enforcing_failure_percentage` is 0. Do we want to preserve that?

Also, as you wrote it, at least one of `success_rate_ejection` or `failure_percentage_ejection` must be set, but that object can contain all `null` fields, which I assume should be functionally equivalent to not setting the outer field at all. Was that your intention?

Reply:
My general concern here is that since we are going to support this for both xDS and non-xDS use-cases, I think we should structure the LB policy config in a way that makes sense for the non-xDS use-case, since that's the one that users will interact directly with.  In the xDS use-case, we'll have machinery to convert from the xDS config to whatever form this LB policy config uses, so it will be fully automated, and we can handle more complexity there if we need to.

So given that, I think we should choose both a structure and a set of defaults that are most useful.  It does seem a little arbitrary to me to enable one ejection algorithm and disable another one by default.  It seems like it might be useful to instead have the user explicitly choose which algorithm(s) they want to use, and then assume reasonable defaults for the parameters for those algorithms.  So, for example, we would not enable success rate unless the message was present, but we would not require the `enforcing_success_rate` to be set since its default is 100.  But we could do the same thing for failure percentage by not enabling it unless the message was present, and then assuming a default of 100 for `enforcing_failure_percentage`.

I'm open to counter-arguments here if there's reason to believe that Envoy's defaults are actually the ones we want.  But I have enough experience with Envoy to not want to *assume* that those defaults are the right ones without thinking it through. :)

Reply:
This seems reasonable to me. We could even allow both fields to be unset to make it easier to create a no-op outlier detection config like you mentioned in your other comment.